File: src/main/java/interview/greedy/GreedyFlorist/GreedyFlorist.scala
package interview.greedy.GreedyFlorist

import java.io._
import java.math._
import java.security._
import java.text._
import java.util._
import java.util.concurrent._
import java.util.function._
import java.util.regex._
import java.util.stream._

object GreedyFloristScala {

  // Complete the getMinimumCost function below.
  def getMinimumCost(k: Int, c: Array[Int]): Int = {
    var result = 0
    val buyers = new scala.collection.mutable.ListBuffer[
      scala.collection.mutable.ListBuffer[Int]
    ]()
    for (i <- 0 until k) {
      buyers.append(new scala.collection.mutable.ListBuffer[Int]())
    }
    val sorted = c.sortWith(_ > _)

    for (i <- 0 until sorted.length by k) {
      for (j <- 0 until k) {
        if (i + j < sorted.length)
          buyers(j).append(sorted(i + j))
      }
    }
    for (i <- 0 until k) {
      for (j <- 0 until (buyers(i).length)) {
        result += buyers(i)(j) * (1 + j)
      }
    }

    result
  }

  def main(args: Array[String]) {
    val stdin = scala.io.StdIn

    val printWriter = new PrintWriter(sys.env("OUTPUT_PATH"))

    val nk = stdin.readLine.split(" ")

    val n = nk(0).trim.toInt

    val k = nk(1).trim.toInt

    val c = stdin.readLine.split(" ").map(_.trim.toInt)
    val minimumCost = getMinimumCost(k, c)

    printWriter.println(minimumCost)

    printWriter.close()
  }
}

File: src/main/java/interview/greedy/GreedyFlorist/greedy_florist.py
#!/bin/python3

# import math
import os

# import random
# import re
# import sys

# Complete the getMinimumCost function below.


def getMinimumCost(k, c):
    result = 0
    buyers = []
    for i in range(k):
        buyers.append([])
    c.sort(reverse=True)
    for i in range(0, len(c), k):
        for j in range(k):
            if i + j < len(c):
                buyers[j].append(c[i + j])
    for i in range(k):
        for j in range(len(buyers[i])):
            result += buyers[i][j] * (1 + j)
    return result


if __name__ == "__main__":
    fptr = open(os.environ["OUTPUT_PATH"], "w")

    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    c = list(map(int, input().rstrip().split()))

    minimumCost = getMinimumCost(k, c)

    fptr.write(str(minimumCost) + "\n")

    fptr.close()
